Extend integration tests to cover inspect option of checkcommitreadiness

satota2 opened the pull request:
#### Type of change

- Test update

#### Description

This patch extends integration tests to cover inspect option of checkcommitreadiness.

#### Related issues

#4428

<!---
#### Release Note
If change impacts current users, uncomment Release Note heading and provide
release note text.
Also, copy release note text into the release specific /release_notes file.
-->

<!--
Checklist (DELETE AFTER READING):

- `Signed-off-by` added to commits (required for DCO check to pass)
- Tests have been added/updated (required for bug fixes and features)
- Unit and/or integration tests pass locally
- Run linters and checks locally using 'make checks'
- If change requires documentation updates, make updates in pull request,
  or open a separate issue and provide link
- Squash commits into a single commit, unless a stack of commits is
  intentional to assist reviewers or to preserve review comments.
- For additional contribution guidelines see the project's CONTRIBUTING.md file
-->

satota2 committed:
Extend integration tests to cover inspect option of checkcommitreadiness

Signed-off-by: Tatsuya Sato <tatsuya.sato.so@hitachi.com>

denyeart commented on integration/nwo/deploy.go:
```diff
@@ -73,8 +73,13 @@ func DeployChaincode(n *Network, channel string, orderer *Orderer, chaincode Cha
 	CheckCommitReadinessUntilReady(n, channel, chaincode, n.PeerOrgs(), peers...)
+
+	// inspect discrepancies between a wrong chaincode definition and each org
+	InspectDiscrepancies(n, channel, chaincode, n.PeerOrgs(), peers...)
```
I'm not following this... if the chaincode was installed and approved for each org, why would there be discrepancies at this step? Could you add some comments explaining it?

C0rWin replied:
IMO, adding an actual test that will utilize this function will demonstrate the intent behind it.

satota2 replied:
@denyeart @C0rWin Thank you for your review and comment.

Aiming to emulate a more realistic case, I have updated the test scenario. As a step prior to making the correct approvals, the test approves a chaincode definition with some incorrect parameters intentionally. Following this, the test utilizes the checkcommitreadiness command with the inspect option to check and demonstrate the ability to identify these discrepancies.

Does this address your concerns?

denyeart commented on integration/nwo/deploy.go:
```diff
@@ -506,6 +514,69 @@ func checkCommitReadiness(n *Network, peer *Peer, channel string, chaincode Chai
+				ChannelID:           channel,
+				Orderer:             n.OrdererAddress(orderer, ListenPort),
+				Name:                chaincode.Name,
+				Version:             "incorrect-version", // Intentionally set incorrect version
```
Much improved, the test now shows the intent of the new inspect option.

I still have a question though... if all the orgs approve the same incorrect version, won't there be no discrepancies to detect? Maybe I'm missing something...

satota2 replied:
@denyeart In this case, I assume that the organization executing commitreadiness has also specified the incorrect parameters in its own approve  mistakenly.
Since the correct chaincode definition is specified in the argument to commitreadiness, all organizations will detect the mismatch.

Other scenario ideas are a case where the organization executing commitreadiness with inspect only specifies the correct version, or where only one organization specifies incorrect version. In these scenarios, the test code may become a bit more complex.

denyeart replied:
Ah right, they all approve the same incorrect definition, but checkcommitreadiness checks for the correct definition. Looks good!

denyeart reviewed (commented):
Ok, I see why other tests are failing. They use the nwo.DeployChaincode() and make assumptions, for example the number of blocks that were required to deploy the chaincode, which is now changed. We don't want to run these extra checks every time a test calls DeployChaincode().

So I'd suggest one of the following:
- Go back to the original PR approach. Now that I understand the discrepancy wasn't between the orgs, it was between the approved chaincode definition and the checkcommitreadiness (incorrect) inputs, I am fine with the approach. I'd just suggest adding some comments to clarify it so that others are not confused.
- Add a new test for the discrepancy instead of leveraging the existing DeployChaincode() function. The downside is that it will extend the overall test execution time since chaincode deployment is the most time consuming aspect across the tests.

satota2 commented:
@denyeart 
Thank you for your feedback. I wasn't aware of the dependency between DeployChaincode() and other tests, and I apologize for the trouble. I have chosen your suggested approach 1 and updated the PR accordingly. I have also added some comments to clarify the scenario, to prevent any confusion for others in the future.I appreciate your review on the updated PR. Thank you.

denyeart reviewed (approved):
Thank you!
